### PR TITLE
add onnx graph-optimization library

### PR DESCRIPTION
**Pull request by anderspapitto**

### CODE REVIEW

File: onnx_caffe2/bin/conversion.py [hunk 1]
```diff
@@ -79,7 +79,8 @@ def caffe2_to_onnx(caffe2_net,
               help='Output path for the caffe2 init net file')
 def onnx_to_caffe2(onnx_model, output, init_net_output):
     onnx_model_proto = ModelProto()
-    onnx_model_proto.ParseFromString(onnx_model.read())
+    onnx_model_proto.ParseFromString(c2.optimize_onnx(onnx_model.read()))
```
Comment (bddppq) [inline onnx_caffe2/bin/conversion.py]: There are two ways of using onnx-caffe2, one is do file to file conversion (here), another is use the python interface to run an onnx model via caffe2 on the fly (which is in backend.py). Both will go through the same translation path `onnx_graph_to_caffe2_net`, so this `optimize_onnx` should be injected into there.

File: manipulation/src/utils/auto_gpu.h [hunk 2]
```diff
@@ -0,0 +1,69 @@
...
+#include <stdexcept>
+
+#include <ATen/ATen.h>
+
+#ifdef WITH_CUDA
+#include <cuda.h>
+#include <cuda_runtime.h>
+#endif
+
+struct AutoGPU {
```
Comment (dzhulgakov) [inline manipulation/src/utils/auto_gpu.h]: we probably don't need gpu access or anything related to it

**Comment by bddppq** [comment]
Please create a "third_party" directory and put all submodules into it.

File: manipulation/src/onnx.pb.cc [hunk 3]
```diff
@@ -0,0 +1,8946 @@
+// Generated by the protocol buffer compiler.  DO NOT EDIT!
```
Comment (dzhulgakov) [inline manipulation/src/onnx.pb.cc]: Having yet another checked-in copy of protobuf is messy. Can we depend on generated protoc files from onnx binary installation or just invoke protoc as a part of cmake script directly?

File: manipulation/src/utils/auto_gpu.h (hunk 2 above)
Comment (anderspapitto) [reply manipulation/src/utils/auto_gpu.h]: sure, I'll go through and clear more of that stuff out

File: manipulation/src/onnx.pb.cc (hunk 3 above)
Comment (anderspapitto) [reply manipulation/src/onnx.pb.cc]: probably - this was the quickest way to get things working, but I'll try one of those approaches

File: onnx_caffe2/bin/conversion.py (hunk 1 above)
Comment (anderspapitto) [reply onnx_caffe2/bin/conversion.py]: that can probably be done, but it requires a more extensive refactoring because

- onnx_graph_to_caffe2_net operates on Graphs, and optimize-onnx needs the whole Model
- process() does destructive things before calling onnx_graph_to_caffe2_net, like stripping out the initializers

**Comment by anderspapitto** [comment]
++ for third_party

**Comment by anderspapitto** [comment]
addressed @jamesr66a comments from the PR in the other repo. DIdn't address the other comments yet.

**Comment by anderspapitto** [comment]
addressed all comments except

- not adding a copy of generated protobuf files
- moving entry point into onnx_graph_to_caffe2

**Comment by anderspapitto** [comment]
protoc now runs during build, and onnx/onnx is submoduled to gain access to the .proto file

**Comment by anderspapitto** [comment]
moved optimize_onnx() into onnx_graph_to_caffe2(), so all comments are now addressed

@bddppq do you actually think this is preferable? I had to substantially shuffle around the innards of backend.py

File: onnx_caffe2/backend.py [hunk 4]
```diff
@@ -648,31 +659,33 @@ def onnx_graph_to_caffe2_net(cls, graph_def):
         for node in graph_def.node:
             predict_net.op.extend(cls._onnx_node_to_caffe2_op(node))
 
-        predict_net.external_input.extend(
-            value_info.name for value_info in graph_def.input)
+        if include_inputs_and_initializers:
```
Comment (bddppq) [inline onnx_caffe2/backend.py]: This part should not be skipped, otherwise the predict_net would be invalid (inner nodes refer to nonexistent input blobs).

File: onnx_caffe2/backend.py [hunk 5]
```diff
@@ -631,9 +639,12 @@ def _all_names_in_graph(graph):
         return names
 
     @classmethod
-    def onnx_graph_to_caffe2_net(cls, graph_def):
+    def onnx_graph_to_caffe2_net(cls, model, include_inputs_and_initializers):
```
Comment (bddppq) [inline onnx_caffe2/backend.py]: rename this, it's not taking graph as input anymore

File: .travis/build.sh [hunk 6]
```diff
@@ -3,10 +3,12 @@
 scripts_dir=$(dirname $(readlink -e "${BASH_SOURCE[0]}"))
 source "$scripts_dir/common";
 
+pip install protobuf PyYAML
+
 [[ -n "$CAFFE2_VERSION" ]] || die "CAFFE2_VERSION not set"
 
 onnx_c2_dir="$PWD"
-pip install -e $onnx_c2_dir
+CC=gcc-5 CXX=g++-5 pip install -e $onnx_c2_dir
```
Comment (dzhulgakov) [inline .travis/build.sh]: why gcc5?

File: manipulation/src/export.cpp [hunk 7]
```diff
@@ -0,0 +1,233 @@
...
+}
+
+const Dimension* dd = nullptr;
+
+void encodeTypeProtoTensorType(onnx::TypeProto_Tensor* tensor_type, Value* n) {
+  onnx::TensorShapeProto* shape = tensor_type->mutable_shape();
+  TensorType* node_type = n->type().get();
+  const std::vector<Dimension>& dims = node_type->sizes();
+  for (const Dimension& d : dims) {
+    dd = &d;
```
Comment (dzhulgakov) [inline manipulation/src/export.cpp]: not sure where it came from, but it's weird

File: manipulation/src/export.cpp [hunk 8]
```diff
@@ -0,0 +1,233 @@
...
+  onnx::TypeProto* t = v->mutable_type();
+  onnx::TypeProto_Tensor* tensor_type = t->mutable_tensor_type();
+  if (n->hasType()) {
+    encodeTypeProtoTensorType(tensor_type, n);
+  }
+}
+
+void encodeGraph(onnx::GraphProto * p_g, const std::shared_ptr<Graph> & g) {
+  JIT_ASSERT(p_g != nullptr);
+  p_g->set_name("torch-jit-export");
```
Comment (dzhulgakov) [inline manipulation/src/export.cpp]: let's preserve the original graph name

File: manipulation/src/test.py [hunk 9]
```diff
@@ -0,0 +1,29 @@
...
+from torch.autograd import Variable
+
+def roundtrip_helper(roundtrip_binary, input_proto_path, output_proto_path):
+    rt = Popen("{} < {} > {}".format(roundtrip_binary, input_proto_path, output_proto_path), shell=True)
+    assert rt.wait() == 0
+
+    input_model = onnx.load(input_proto_path)
+    output_model = onnx.load(output_proto_path)
+
+    assert onnx.helper.printable_graph(input_model.graph) \
```
Comment (dzhulgakov) [inline manipulation/src/test.py]: it might be better to compare source protobufs (e.g. str(graph_a) == str(graph_b)) directly as printable_graph() doesn't print all fields.

File: manipulation/src/test.py [hunk 10]
```diff
@@ -0,0 +1,29 @@
...
+        == onnx.helper.printable_graph(output_model.graph)
+
+def test_torch_model(roundtrip_binary):
```
Comment (dzhulgakov) [inline manipulation/src/test.py]: not invoked?

File: manipulation/src/test.py [hunk 11]
```diff
@@ -0,0 +1,29 @@
...
+    model=torch.nn.Sequential(torch.nn.Linear(3,2), torch.nn.Linear(2,1))
+    x = Variable(torch.randn(10, 3))
+    input_path = "/tmp/model.onnx"
+    output_path = "/tmp/out.onnx"
+    torch.onnx.export(model, x, input_path)
+    roundtrip_helper(roundtrip_binary, input_path, output_path)
+
+# invoke as
```
Comment (dzhulgakov) [inline manipulation/src/test.py]: maybe turn it to unittest and invoke some of the models from the model zoo?

File: manipulation/src/main.cpp [hunk 12]
```diff
@@ -0,0 +1,21 @@
...
+// doc_strings are lost, but otherwise should be unchanged.
+int main(int argc, char** argv) {
+  std::stringstream buffer;
+  buffer << std::cin.rdbuf();
+
+  onnx::ModelProto mp;
+  mp.ParseFromString(buffer.str());
+  std::shared_ptr<onnx::optimization::Graph> g = onnx::optimization::ImportModel(mp);
+
+  std::string s = onnx::optimization::ExportGraph(g);
```
Comment (dzhulgakov) [inline manipulation/src/main.cpp]: you probably want to preserve original metadata from ModelProto

File: onnx_caffe2/backend.py [hunk 13]
```diff
@@ -472,6 +473,12 @@ def _direct_initialize_inputs(cls, inputs, initialized, ws, device_option):
             shape = list(d.dim_value for d in value_info.type.tensor_type.shape.dim)
             ws.FeedBlob(value_info.name, np.ones(shape), device_option)
 
+    @staticmethod
+    def optimize_onnx(input):
+        proc = Popen(['optimize-onnx'], stdin=PIPE, stdout=PIPE, stderr=PIPE)
```
Comment (dzhulgakov) [inline onnx_caffe2/backend.py]: should it be a relative path?

File: onnx_caffe2/backend.py [hunk 14]
```diff
@@ -629,10 +624,14 @@ def _all_names_in_graph(graph):
         return names
 
     @classmethod
-    def onnx_graph_to_caffe2_net(cls, graph_def, device="CPU"):
+    def onnx_model_to_caffe2_net(cls, model, include_inputs_and_initializers, device="CPU"):
```
Comment (dzhulgakov) [inline onnx_caffe2/backend.py]: make it the last argument (as some callsites might pass device already) and default to true

File: .travis/build.sh (hunk 6 above)
Comment (bddppq) [reply .travis/build.sh]: @dzhulgakov pytorch gcc4.8 build is currently broken.
@anderspapitto you probably also need to configure travis to install gcc5

File: onnx_caffe2/backend.py (hunk 14 above)
Comment (bddppq) [reply onnx_caffe2/backend.py]: the function name has been changed, this will be a breaking change anyway (but to be fair this function is never meant to be a public function)

File: setup.py [hunk 15]
```diff
@@ -57,15 +59,38 @@ def run(self):
...
+        self.run_command('build')
         setuptools.command.develop.develop.run(self)
+        source = os.path.join(os.getcwd(), 'build/manipulation/src/optimize-onnx')
+        target = os.path.join(sys.prefix, 'bin/')
+        subprocess.check_call(['ln', '-s', '-f', source, target])
+
+class install(setuptools.command.install.install):
+    def run(self):
+        setuptools.command.install.install.run(self)
+        source = os.path.join(os.getcwd(), 'build/manipulation/src/optimize-onnx')
```
Comment (bddppq) [inline setup.py]: for the purpose of cross platform, use `os.path.join` to concat each component

File: setup.py [hunk 16]
```diff
@@ -57,15 +59,38 @@ def run(self):
...
+    def run(self):
+        self.run_command('build_py')
+        build_optimize_onnx()
+
 class develop(setuptools.command.develop.develop):
     def run(self):
         self.run_command('create_version')
+        self.run_command('build')
         setuptools.command.develop.develop.run(self)
+        source = os.path.join(os.getcwd(), 'build/manipulation/src/optimize-onnx')
```
Comment (bddppq) [inline setup.py]: Probably don't rely on `getcwd`, use `__file__`?

File: manipulation/src/test.py (hunk 9 above)
Comment (bddppq) [reply manipulation/src/test.py]: I think you can directly do `graph_a == graph_b`

File: onnx_caffe2/backend.py (hunk 14 above)
Comment (dzhulgakov) [reply onnx_caffe2/backend.py]: it'd actually invoked (we should use '_' prefix for private stuff): https://fburl.com/8z9yn2ck and https://github.com/ezyang/onnx-pytorch/search?utf8=%E2%9C%93&q=onnx_graph_to_caffe2_net&type=

File: manipulation/src/test.py (hunk 9 above)
Comment (anderspapitto) [reply manipulation/src/test.py]: I can do this, but I'll have to add more code to preserve comments everywhere (almost every onnx struct has a comment field that I'm currently not preserving)

File: manipulation/src/test.py (hunk 11 above)
Comment (anderspapitto) [reply manipulation/src/test.py]: ++

File: onnx_caffe2/backend.py (hunk 13 above)
Comment (anderspapitto) [reply onnx_caffe2/backend.py]: I've changed setup.py to put `optimize-onnx` in the bin/ of the python install/virtualenv, so it should be fine to look for it in PATH

File: setup.py (hunk 16 above)
Comment (anderspapitto) [reply setup.py]: ++

File: .travis/build.sh (hunk 6 above)
Comment (anderspapitto) [reply .travis/build.sh]: yeah, I have modified the travis setup in this diff.

(specifically, it's aten which doesn't build with gcc4.8

File: manipulation/src/export.cpp (hunk 7 above)
Comment (anderspapitto) [reply manipulation/src/export.cpp]: ++

File: onnx_caffe2/backend.py (hunk 13 above)
Comment (bddppq) [reply onnx_caffe2/backend.py]: @anderspapitto This is not always true, you can use the virtualenv without activating the virtualenv (thus `virtualenv/bin` is not in your `$PATH`).

File: tests/roundtrip_optimize_onnx.py [hunk 17]
```diff
@@ -0,0 +1,67 @@
...
+        print (onnx.helper.printable_graph(before_roundtrip.graph))
+        print('\n\nAFTER\n')
+        print (onnx.helper.printable_graph(after_roundtrip.graph))
+
+        assert onnx.helper.printable_graph(before_roundtrip.graph) \
+            == onnx.helper.printable_graph(after_roundtrip.graph)
+
+    # arbitrarily pick one relatively small model to sanity test with
+    def test_squeezenet(self):
+        self._roundtrip('squeezenet-ir-version-3')
```
Comment (anderspapitto) [inline tests/roundtrip_optimize_onnx.py]: note: this passes locally, but won't pass in CI until I'm able to upload the new ir version of this model

File: optimize-onnx/src/main.cpp [hunk 18]
```diff
@@ -0,0 +1,58 @@
...
+int main(int argc, char** argv) {
+  std::stringstream buffer;
+  buffer << std::cin.rdbuf();
+
+  onnx::ModelProto mp_in;
+  mp_in.ParseFromString(buffer.str());
+  std::shared_ptr<onnx::optimization::Graph> g = onnx::optimization::ImportModel(mp_in);
+
+  if (g.get() == nullptr) {
+    std::cout << buffer.str();
```
Comment (dzhulgakov) [inline optimize-onnx/src/main.cpp]: print some warning to stderr?

File: optimize-onnx/src/main.cpp [hunk 19]
```diff
@@ -0,0 +1,58 @@
...
+
+#include "export.h"
+#include "import.h"
+
+
+// reads a protobuf file on stdin and writes it back to stdout.
+// doc_strings are lost, but semantics should be unchanged.
+int main(int argc, char** argv) {
+  std::stringstream buffer;
+  buffer << std::cin.rdbuf();
```
Comment (dzhulgakov) [inline optimize-onnx/src/main.cpp]: this might a problem with reading binary data: https://stackoverflow.com/questions/7587595/read-binary-data-from-stdcin or does rdbuf() solve all of that?

File: onnx_caffe2/backend.py [hunk 20]
```diff
@@ -627,10 +631,23 @@ def _all_names_in_graph(graph):
         return names
 
     @classmethod
-    def onnx_graph_to_caffe2_net(cls, graph_def, device="CPU"):
+    def _onnx_model_to_caffe2_net(cls, model, device, include_inputs_and_initializers):
         device_option = get_device_option(Device(device))
+
+        # before_opt = model.SerializeToString()
```
Comment (dzhulgakov) [inline onnx_caffe2/backend.py]: remove?

**Comment by bddppq** [comment]
We haven't done any binary distribution for onnx packages so far. With the addition of c++ component, bdist would definitely be more difficult, let's just use source distribution for now.

**Comment by ezyang** [comment]
We have conda-forge packaging for onnx-caffe2 now (https://github.com/conda-forge/onnx-caffe2-feedstock) so binary distribution should be easy.

File: onnx_caffe2/backend.py (hunk 20 above)
Comment (anderspapitto) [reply onnx_caffe2/backend.py]: ++

File: optimize-onnx/src/import.cpp [hunk 21]
```diff
@@ -0,0 +1,252 @@
...
+      // we don't know the type here, so that's done in a later pass
+      out->setUniqueName(np.output(j));
+      value_by_name_of[np.output(j)] = out;
+    }
+    convertAttributes(np, n);
+    std::vector<std::string> inputs;
+    inputs.reserve(np.input_size());
+    for (int i = 0; i < np.input_size(); i++) {
+      inputs.push_back(np.input(i));
+      inputs_by_node[n] = inputs;
```
[the remark this answers is not in the document]
Comment (anderspapitto) [reply optimize-onnx/src/import.cpp]: ++

**Comment by anderspapitto** [comment]
Yes, I need to rebase and finish fixing one (known) bug (I always rely on raw_data, rather than float_data/int_data/string_data, etc).

As far as I know it should not be a big problem for binary distributions, unless they have difficulty with native executables. I can test that it's compatible with whatever solutions we currently have (e.g. the conda package).

**Comment by anderspapitto** [comment]
rebased, added support for all the various <datatype>_data fields in TensorProto, and along the way largely removed the ATen dependency

File: onnx_caffe2/backend.py [hunk 22]
```diff
@@ -501,6 +504,20 @@ def _direct_initialize_inputs(cls, inputs, initialized, ws, device_option):
             shape = list(d.dim_value for d in value_info.type.tensor_type.shape.dim)
             ws.FeedBlob(value_info.name, np.ones(shape), device_option)
 
+    @staticmethod
+    def optimize_onnx(input):
+        # this is where the binary is located when onnx-caffe2 has been installed
```
Comment (dzhulgakov) [inline onnx_caffe2/backend.py]: you might need to modify it for internal BUCK-based build. Probably, try to import optimizer as a python module if it's installed and otherwise - fallback on the binary.

Comment (anderspapitto) [reply onnx_caffe2/backend.py]: ok, I'll be aware of that for when it comes up

**Comment by anderspapitto** [comment]
now that it's approved, I've rebased and squashed most of the commits down. Assuming tests still pass someone with write access can merge it.
